src/error.rs

```diff
@@ -285,6 +290,36 @@ impl StdError for ValidateError {
     }
 }
 
+/// Tag was corrupted during session.
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct TagCorrupted {
+    /// Expected tag number
+    pub(crate) expect: u32,
+    /// Actual tag number, 0 if parse failed
+    pub(crate) actual: u32,
+}
+
+impl fmt::Display for TagCorrupted {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Expected tag number is {}, actual {}",
+            self.expect, self.actual
+        )
+    }
+}
+
+impl StdError for TagCorrupted {
+    fn description(&self) -> &str {
+        "Tag is corrupted, session is in an inconsistent state"
+    }
+
+    fn cause(&self) -> Option<&dyn StdError> {
+        None
+    }
```

Comment:
I think you can skip both of these `fn`s — they're not required and their default implementations will do the right thing.